### ISSUE TITLE
selinux_policy_port doesn't correctly detect if existing mappings exist

### ISSUE THREAD

**Issue by datallah** [BackSlasher/chef-selinuxpolicy#30]
When a secontext has multiple ports, the not_if/only_if commands don't recognize it correctly:
```bash
/usr/sbin/semanage port -l | grep -P 'http_cache_port_t'
http_cache_port_t              tcp      3128, 8080, 8118, 8123, 10001-10010
http_cache_port_t              udp      3130

#When using the following resource
selinux_policy_port 8080 do
  protocol 'tcp'
  secontext 'http_cache_port_t'
end
#This is what the not_if looks like
/usr/sbin/semanage port -l | grep -P 'http_cache_port_t\s+tcp\s+8080'
#(no matches)
#It needs to look something like this:
/usr/sbin/semanage port -l | grep -P 'http_cache_port_t\s+tcp\s+([0-9-]+,\s+)*8080(,\s+[0-9-]+)*$'
```
It'll also mistakenly do partial matches:
```
/usr/sbin/semanage port -l | grep -P 'http_cache_port_t\s+tcp\s+312'
http_cache_port_t              tcp      3128, 8080, 8118, 8123, 10001-10010
```

**Comment by datallah** [comment]
I guess this is what pull #12 is referring to as well, except that the patch there doesn't look like it's complete.

**Comment by BackSlasher** [comment]
OK, after thinking about it for a while, I want to leave the entire regex detection idea behind. However you tweak it, you can't use it to detect ranges (e.g. that 81 is in `80-82`).
See commit c0d67cea453375a2882c5da70eafb1434568c6af. It is no way tested as I have other stuff to attend to :(, but if any of you skilled guys want to make things better, that's the direction I'd love to see.
The basic idea is moving to python-based guards (and not shell ones), and use the `seboject` python module directly to look for that port. We can later upgrade the resource to support ranges, but that's low priority.

I'd also appreciate some testing, since we have some crude framework

Record keeping: #12 #27 #29 
CC @Sauraus @equick

**Comment by datallah** [comment]
I haven't actually tested c0d67ce, but I see a couple issues:
 * I don't think it's actually filtering by port anywhere
 * it doesn't handle the "if" block within the addormodify action - that's still with a grep

It seems like it'd be better to make this work using [load-current-resource] (https://docs.chef.io/lwrp_custom_provider.html#load-current-resource) and to look up the entire selinux state once and cache it in memory instead of reading the state every time it's needed.

**Comment by BackSlasher** [comment]
Yeah it's very preliminary. I'm still toying around with that.
How will `load-current-resource` help with loading the entire state? Are you suggesting some sort of singleton cache dictionary thingy?

**Comment by datallah** [comment]
load-current-resource won't necessarily help with the caching state, but it's a more chef native approach to implementing resources.

Yes, I was thinking there'd be some sort of singleton cache (and it likely should get invalidated/updated if any of the resources actually make changes).

**Comment by BackSlasher** [comment]
`load-current-resource` sure makes it prettier, but more complicated as well. I see no urgent need to implement it, but PRs are welcome :)
About the singleton cache - this is a bad idea IMO because the SELinux policy can be modified by other things than this cookbook, like custom scripts, packages being installed, etc. I can barely rely on the system being stable when the resource's action is being executed.

**Comment by bdclark** [comment]
Here's a possible pattern using python and `seobject`, but instead of using python-based guards it uses a helper and shell_out: https://github.com/bdclark/chef-selinuxpolicy/commit/bfc3fd42e540bb9b1ac44cd5d144e272e946ac0d.

Seems to match ports, protocol, etc. correctly but honestly it's only lightly tested at this point (against the existing TK suite). Just wanted to throw it out there as an idea.

**Comment by BackSlasher** [comment]
I made a second attempt [here](/backslasher/chef-selinuxpolicy/tree/python-ports2)
The earlier commit was complete rubbish - that's what you get when writing stuff on the bus.
Anyway, the kitchen tests went fine. Since I don't actually use the port functionality, can someone who does give his opinion please?

**Comment by BackSlasher** [comment]
Should be resolved with 06862589c15b7ca8924131231db023cb33b1ab1d. At least my tests tell me so.  
If it's still an issue, post here please
